layers/ifa.go

```diff
@@ -0,0 +1,353 @@
+package layers
+
+import (
+	"encoding/binary"
+	"fmt"
+
+	"github.com/google/gopacket"
+)
+
+const (
+	// IFA version
+	IFAVersion uint8 = 0x02
+	// GNS disabled, metadata in an IFA zone is defined by
+	// the LNS of each hop
+	NoGlobalNameSpace = 0x0f
+)
+
+// IFAPortSpeed interface speed used to specified the egress
+// interface port speed in IFA.
+// Encodings are 0–10Gbps, 1–25Gbps, 2–40Gbps, 3–50Gbps,
+// 4–100Gbps, 5–200Gbps, 6–400Gbps.
+type IFAPortSpeed uint8
+
+const (
+	IFAPortSpeed10G IFAPortSpeed = iota
+	IFAPortSpeed25G
+	IFAPortSpeed40G
+	IFAPortSpeed100G
+	IFAPortSpeed200G
+)
+
+// Register IFA layer
+func decodeIFA(data []byte, p gopacket.PacketBuilder) error {
+	ifa := &IFA{}
+
+	err := ifa.DecodeFromBytes(data, p)
+	if err != nil {
+		return err
+	}
+	p.AddLayer(ifa)
+	// TODO: should we use the application layer
+	//       and create two layer type,
+	//       one for header and one for metadata
+	p.SetApplicationLayer(ifa)
+
+	return nil
+}
+
+/*
+IFA header is inserted between layer 3 and 4 by the initiator node. Then
+Metadata header, fragment, checksum and data are inserted between layer
+4 and payload or at the end of the packet just before the FCS.
+Each node append metadata to the stack and terminating node remove all
+IFA headers/data from the packet, forward it and generates and sends a
+report to am IPFIX collector.
+
```

Comment:
I assume the report to the IPFIX collector includes packet samples with 

1. IFA Header
2. IFA Metadata stack & header

otherwise we would not be adding this to the package, right?

Reply:
yes there is 3 information elements in the IPFIX report:

- IFA headers
- IFA metadatas
- first packet bytes (128 for broadcom, not configurable)